### ISSUE TITLE
Improve the behaviour of \newline and related commands

### ISSUE THREAD

**Issue by car222222** [latex3/latex2e#627]
## Brief outline of the enhancement

Extend the usability of the \newline and similar commands immediately after either a run-in heading or a list item.

## Notes

The example below does not cover all cases.
A good fix is to add code such as the following, much as us done within \newpage:
\ifvmode
   \if@noskipsec 
      \leavevmode 
      \global \@noskipsecfalse
   \fi 
   \if@inlabel 
      \leavevmode 
      \global \@inlabelfalse 
   \fi
\fi

## Minimal example showing the current behaviour

```
\RequirePackage{latexbug}       % <--should be always the first
\documentclass{article}
\begin{document}
\section{Section title}
%%  Maybe need to add some subsections etc here? 
%%  Bug is that both the \\ here produce ‘nolnerr’ due to (not obviously) 
%%    being in vmode.  
%%  Fix is to add a \ leavevmode in these (and several similar) cases.
\subparagraph*{This is a long tedious text for this run-in subparagraph heading} \\
A list.
\begin{description}
\item[Boring and overlong descriptive text) \\
My interesting label body text goes here. 
\end {description}
\end{document}
```

## Minimal example showing the desired new behaviour

**Comment by blefloch** [comment]
More generally it would be useful to have a command defined as `\ifvmode \if@noskipsec \leavevmode \global @noskipsecfalse \fi \if@inlabel \leavevmode \global \@inlabelfalse \fi \fi` or similar, meant to "put here everything that has been deferred".  This could be useful in situations like `\begin{itemize} \item foo \item \color{red} bar \end{itemize}` where the second bullet point is surprisingly colored.

**Comment by car222222** [comment]
Good dea!  I too was thinking along those lines. 

But need to check if exactly the same logic is correct in all such situations.

**Comment by car222222** [comment]
Yes, this code (plus possibly other tests?) could usefully be consolidated in a command, maybe called:  \check@leavevmode ??

**Comment by car222222** [comment]
But making assignments (parameter setting), such as setting the text color, font size etc., cannot in general be handled in this way.

It looks like a good use of the new hook ideas, to get the assignments executed in the right place.

Thus it needs a separate issue.

**Comment by car222222** [comment]
The original post in this issue was a simple code adjustment, very close to being just a minor bug fix.

It was submitted as an easy fix in response to one of a long list of "problems" that @u-fischer presented some time ago. @blefloch widely expanded the scope.

The original minor fix can easily be applied now, so maybe
 @u-fischer should restate her original request, together with the suggested fix, in a new issue.  This would remove the confusion that has led to this issue being marked as long-term.
